### DISCUSSION TITLE
KeyChain MAC/iOS

### DISCUSSION THREAD

**Discussion by angelru** [dotnet/maui#6879] [Q&A]
in xamarin I can access the keychain through `SecRecord`, in MAUI I don't see anything...

**Comment by nor0x** [comment]
You can use `SecureStorage`  it's an abstraction over secure storages on various platforms. Previously it was available in Xamarin.Essentials but it's now built-in to MAUI.  On iOS and macOS it uses keychain to store values:
`https://github.com/dotnet/maui/blob/main/src/Essentials/src/SecureStorage/SecureStorage.ios.tvos.watchos.macos.cs`

```
await SecureStorage.SetAsync("secret-key", "secret");
var secret= await SecureStorage.GetAsync("secret-key");
```

afaik the documentation is still valid:
https://docs.microsoft.com/en-us/xamarin/essentials/secure-storage?tabs=ios

**Reply by angelru** [reply]
> You can use `SecureStorage` it's an abstraction over secure storages on various platforms. Previously it was available in Xamarin.Essentials but it's now built-in to MAUI. On iOS and macOS it uses keychain to store values: `https://github.com/dotnet/maui/blob/main/src/Essentials/src/SecureStorage/SecureStorage.ios.tvos.watchos.macos.cs`
> 
> ```
> await SecureStorage.SetAsync("secret-key", "secret");
> var secret= await SecureStorage.GetAsync("secret-key");
> ```
> 
> afaik the documentation is still valid: https://docs.microsoft.com/en-us/xamarin/essentials/secure-storage?tabs=ios

I mean to recover the certificates installed on a MAC/iOS

```
    //        SecRecord rec = new SecRecord(SecKind.Certificate);
    //        SecStatusCode code;
    //        SecKeyChain.QueryAsRecord(rec, out code);
    //        NSData[] certs = SecKeyChain.QueryAsData(rec, false, 999, out code);
```

**Reply by angelru** [reply]
@nor0x any idea why the code above doesn't return certs on macOS? I have them installed (pfx)

**Comment by Sgudipu** [comment]
@angelru any solution for this get the access from keychain for Mac Catalyst.

**Reply by angelru** [reply]
@Sgudipu I haven't tried it in a while, maccatalyst can't seem to access the keychain, I had to go back to the old xamarin macos
